### DISCUSSION TITLE
Issue with sparse matrix RCM reordering

### DISCUSSION THREAD

**Discussion by rafatcampos** [ginkgo-project/ginkgo#939] [Q&A] [answered]
Cheers!

I have an issue when using the Ginkgo RCM reordering feature. I am using the following configuration, which I got the idea from examples and the documentation:

`    auto rcm_obj = gko::reorder::Rcm<float>::build().with_construct_inverse_permutation(true).on(master)->generate(hA);`
`    auto permute_mtx = rcm_obj->get_permutation();`
`    permute_mtx->apply(lend(A), lend(A_rcm));`

where `A` and `A_rcm` are the original and reordered matrices, respectively. However, when I export the reordered matrix to a MM file to visualize it, it does not correspond to the expected with RCM reordering. Is there anything that should be done differently? I haven't found any examples directly about reordering or RCM in the repository or Gingko website.

Thank you in advance for the help!

**Comment by upsj** [comment, marked as answer]
Thanks for the report! Which executor are you using as `master`, Reference or OpenMP? Can you print the permutation indices and export them together with the generated `.mtx` files to share them here?

**Reply by rafatcampos** [reply]
Hi!

I am using the reference executor as master. I have attached the 2 .mtx files from the original and reordered matrix (garon2 from suite sparse). As for the permutation indices, I am not able to print `permute_mtx`, as it is not the correct data type for `gko::write`. I was not able to find a way to convert it to another format. This is also why I have used the `apply` method to perform reordering instead of the `permute` method with the permutation indices.

[matrices.zip](https://github.com/ginkgo-project/ginkgo/files/7735703/matrices.zip)

**Reply by upsj** [reply]
Thanks for the details, that does look weird, we'll take a look at it!

**Reply by upsj** [reply]
I see the issue: You are permuting the rows only, when you would need to permute rows and columns at the same time. This is somewhat of a design issue we need to address on our end, a working version would be
```
auto perm_array = gko::Array<int>::view(exec, permute_mtx->get_permutation_size(), const_cast<int*>(permute_mtx->get_const_permutation()));
auto A_rcm = gko::as<gko::Csr<...>>(A->permute(&perm_array));
```

**Reply by rafatcampos** [reply]
Thank you @upsj ! Your fix was able to solve the issue entirely! RCM was doing its work all along, the problem was in the permute part. Thanks again for the support and great work on Ginkgo!
